Enable to serialize null relations closes #10

jo-tham opened the pull request:
Hi @sloria 

this is neither of the PR's I've promised you lately... but this one was more urgent for me :wink: 

I testing this out in a flask application, and the null relationship fields in the json are simply missing. So while it seems this patch is OK, the desired result for a schema dump is not complete.

I think we need to add a test to `test_schema.py` for serializing relatonships, including null relationship and then deal with why the objects are not being serialized properly with null relationship. 

I looked around, seems like marshmallow-jsonapi.schema.format_item() does some magic with dict.update, which I think may be the cause. I suspect using {} instead of the jsonapi standards for empty relationships might be work with the given format_item, looks like it might give an empty collection for the relationship.

Any suggestions?

jo-tham committed:
Enable to serialize null relations closes #10

Not very DRY and I'm not familiar with pytest. But this should close #10

for the relationship, jsonapi requires 'null' for an empty to-one
relationship or '[]' for an empty to-many relationship. This caused some
diversion from the discussion in #10 which suggested returning '{}' for
any case.

http://jsonapi.org/format/#document-resource-object-linkage

jo-tham commented:
hm, I added some tests for the serializer; it seems to work OK for null relationships. Let me know if you want me to push them up; they may be redundant. Probably a problem with my app, then...

jo-tham commented:
yep, it was in a issue in how i defined relationship in the schema. actually, the schema relationship attribute had to use the same name as the relationship attribute of the object being serialized. this generally applies but somehow I have not noticed it until now.

sloria commented:
According to JSON API 1.0:

```
A "relationship object" MUST contain at least one of the following:

links: a links object containing at least one of the following:
self: a link for the relationship itself (a "relationship link"). This link allows the client to directly manipulate the relationship. For example, it would allow a client to remove an author from an article without deleting the people resource itself.
related: a related resource link
data: resource linkage
meta: a meta object that contains non-standard meta-information about the relationship.
```

http://jsonapi.org/format/1.0/#document-resource-object-relationships

So it seems that null relationships should be omitted from responses rather than serializing to `null`.

jo-tham commented:
Good catch, I was looking at compound documents here

http://jsonapi.org/format/1.0/#document-resource-object-linkage

I guess I have some concern with jsonapi, then. At the risk of going outside the scope here what are your thoughts/knowledge of rest API best practice with respect to this:
Should't the response define the schema? if we omit an empty relationship, it is not possible for client to know if relationship exists and is missing, or there is no such relationship.

Reading further

```
A server MUST respond to a successful request to fetch an individual resource with a resource object or null provided as the response document's primary data.

null is only an appropriate response when the requested URL is one that might correspond to a single resource, but doesn't currently.

Note: Consider, for example, a request to fetch a to-one related resource link. This request would respond with null when the relationship is empty (such that the link is corresponding to no resources) but with the single related resource's resource object otherwise.
```

This seems to imply that there should be a link for a related resource, which resolves to none if the related resource doesn't exist. _Then_ How do we get an id to create a link for a nonexistent resource?

edit: the related resource is at, e.g. , posts/1/comments or posts/1/author which may not exist, but a request to them doesn't need the id of the related object; it returns empty data

sloria commented:
The section you are referencing is referring to _primary data_, not relationship objects.

According to http://jsonapi.org/format/1.0/#document-resource-object-linkage, it seems the correct place to represent an empty/null relationship is in a resource linkage, e.g.

``` JSON
{
    "data": {
        "id": "abc12",
        "type": "posts",
        "relationships": {
            "author": {
                "data": null
            }
        }
    }
}
```

jo-tham commented:
Ageed. 

To elaborate... Verbatim from http://jsonapi.org/format/1.0/#document-resource-object-linkage

```
Resource Linkage

Resource linkage in a compound document allows a client to link together all of the included resource objects without having to GET any URLs via links.

Resource linkage MUST be represented as one of the following:

    null for empty to-one relationships.
    an empty array ([]) for empty to-many relationships.
    ...
```

so a "**compound**" (see also #11) document should represent resource _linkages_ in the format that this pull request achieves (pending your verification!). 

It is not clear to me how these linkages should be represented in _non_-compound documents. My interpretation is that the above rules for null/empty resource linkages in _non_-compound documents, too. But I can verify with the jsonapi crew...

jo-tham commented:
updated to return literal null instead of string for empty to-one relationship

sloria commented:
After reading [here](http://jsonapi.org/format/#document-resource-object-relationships), [here](http://jsonapi.org/format/#document-resource-object-linkage) and glancing at the [JSON Schema](http://jsonapi.org/schema) again, it appears that the following is valid:

``` JSON
"relationships": {
    "parent": {
        "data": null
    }
}
```

because `/relationships/parent/data` is a [resource linkage](http://jsonapi.org/format/#document-resource-object-linkage), and resource linkages may be `null`.

However, this is invalid:

``` JSON
"relationships": {
    "parent": null
}
```

So it seems that the correct way to serialize null relationships is to have a `null` resource linkage. I'm not sure what the behavior should be for a null relationship if `include_data=False`. I suppose we could serialize to `missing` so that the relationship object would be non-existent.

jo-tham commented:
I think you are correct - good eye!

For the record

```
A "relationship object" MUST contain at least one of the following:
...
    data: resource linkage
...
```

And, regardless of compound document or not

```
Resource linkage MUST be represented as one of the following:

    null for empty to-one relationships.
    an empty array ([]) for empty to-many relationships.
    a single resource identifier object for non-empty to-one relationships.
    an array of resource identifier objects for non-empty to-many relationships.
```

As for `include_data=False`

if we don't include data, we must have 'links' or 'meta' - http://jsonapi.org/format/#document-resource-object-relationships. I think we should have a link for the relationship. The implementation (outside of marshmallow-jsonapi scope) should respond to GET with [] or null. See below

```
A server MUST respond to a successful request to fetch a resource collection with an array of resource objects or an empty array ([]) as the response document's primary data.

...

A server MUST respond to a successful request to fetch an individual resource with a resource object or null provided as the response document's primary data.

null is only an appropriate response when the requested URL is one that might correspond to a single resource, but doesn't currently.

Note: Consider, for example, a request to fetch a to-one related resource link. This request would respond with null when the relationship is empty (such that the link is corresponding to no resources) but with the single related resource's resource object otherwise.
```

jo-tham committed:
Fixup missing/empty relationships

If include_data=False we must return meta or links. This caused a change
to a test which predates this commit that had a related url which is
does not conform to jsonapi standard. I.e. the id of the related
resource can not be relied upon.

And now the related *resource linkage*, i.e. 'data' attribute of the
related resource object resolves to null or []

Also added test for include_data=False

jo-tham commented:
I think the last commit gives the desired behaviour. But there's a few places in the docs that should be fixed up for users.

we cannot rely on existence of related objects, and therefore should always define urls in terms of the current object. developers must implement those routes to conform to jsonapi.

``` python
    author = fields.Relationship(
        self_url='/articles/{article_id}/relationships/author',
        self_url_kwargs={'article_id': '<id>'},
        #related_url='/authors/{author_id}',
        #related_url_kwargs={'author_id': '<author.id>'}
        related_url='artciles/{article_id}/author',
        related_url_kwargs={'artcile_id': '<id>'}
    )
```

Ii think this is the case throughout the jsonapi specification
http://jsonapi.org/format/http://jsonapi.org/format/

please confirm.

Let me know if that can be a separate PR...

jo-tham commented:
Ah bollox, I guess this will breaks some of the convenience of the flask related view functionality. Or rather, related views cannot be the basic views on related objects

if we have views for
- posts
- authors

we'll have to add views for
- post_author
- author_posts

Lots of boilerplate.

dgeb commented:
Hi all ... I'm an author of the spec and saw @jo-tham's call for clarification on twitter. I can confirm that @sloria's [comment](https://github.com/marshmallow-code/marshmallow-jsonapi/pull/14#issuecomment-156655270) above is correct. 

Do you have further questions?

jo-tham commented:
@dgeb thanks for the confirmation! Nothing further from me at the moment.

jo-tham commented:
to add to my [last comment](https://github.com/marshmallow-code/marshmallow-jsonapi/pull/14#issuecomment-156866260)

I've done this in some of my work (i.e. implemented these jsonapi spec routes for relationships) and it gives the desired result. it is possible for a marshmallow-jsonapi user to factor out some of the repitition in their (e.g. flask) application

I think that completes this PR. @sloria anything else I should do here?

sloria commented:
@jo-tham Thanks for your work on this.

It would be good to also have tests for serializing null and empty relationships when `include_data=False`. In both these cases, it might make sense to serialize to `marshmallow.missing`, so that the field is excluded from the response.

jo-tham commented:
Should be able to update tomorrow.

jo-tham commented:
Hey @sloria I want to clarify

If include_data = False are you suggesting we omit the entire _relationship_ field or just the data member?

If it's just the data member, we can use a test such as

```
    with pytest.raises(KeyError):
        result['comments']['data']
```

such a test applies regardless of whether the relationship is populated, but I can add for this PR as it doesn't seem to exist.

Since `data` is not a field itself (according to my understanding), we can't set it to `marshmallow.missing`.

I don't think the entire relationship field should be excluded from the response when it is null. although jsonapi seems to allow this, which would constitute omitting an entire relationship object, api client doesn't know if response should have a relationship or not
http://jsonapi.org/format/#document-resource-object-relationships
@dgeb any comment on this?

So i think we should just omit the data field, as currently implemented

sloria commented:
If I'm interpreting the spec correctly, the relationship object should be omitted for both _null and empty relationships_ if a linkage is not requested (i.e. if `include_data=False`). If a linkage is requested, then relationship object will be `{"data": null}`.

In sum:

Post with no author and no comments + linkage requested (`include_data=True`)

``` JSON
{
    "relationships": {}
}
```

Post with no author and no comments + linkage not requested

``` JSON
{
    "relationships": {
        "author": {"data": null},
        "comments": {"data": []}
    }
}
```

@dgeb Can you confirm?

dgeb commented:
I can confirm that the following would represent a post with no author or comments:

```
{
    "relationships": {
        "author": {"data": null},
        "comments": {"data": []}
    }
}
```

Entire relationships should NOT be omitted to represent a lack of data, like this:

```
{
    "relationships": {}
}
```

Relationship objects may contain details, such as a `links` object with `self` and `related`. These links, or a `meta` object, could be present regardless of the current contents of `data`.

Note that the `data` member is only strictly required to fulfill the [full linkage](http://jsonapi.org/format/#document-compound-documents) requirement in compound documents. Besides that, `data` can be included as you wish, although I strongly recommend being consistent. In other words, don't leave off `data` just to indicate that relationships are empty.

sloria commented:
Thanks @dgeb . What should be returned for null/empty relationships if linkages are not supported by an API?

dgeb commented:
> What should be returned for null/empty relationships if linkages are not supported by an API?

If an API supports compound documents, then it needs to support including linkage `data` for relationships to the included resources. Otherwise, there's no obligation to include relationship `data`. Regardless, if your API does include relationship `data`, then empty relationships should be represented as either `data: null` (to-one) or `data: []` (to-many).

sloria commented:
I'm still unclear on what should be returned by an API that does _not_ support compound documents (and therefore doesn't include relationship `data` in the common case).

Omit the relationship objects?

``` json
{
  "relationships": {}
}
```

Include `data` just for null/empty relationships?

``` json
{
  "relationships": {"author": {"data": null}, "comments": {"data": []}}
}
```

Empty relationship objects?

``` json
{
  "relationships": {"author": {}, "comments": {}}
}
```

dgeb commented:
Each [relationship object](http://jsonapi.org/format/#document-resource-object-relationships) MUST contain at least one of the following: `links`, `data`, or `meta`.

Therefore, this is not allowed:

```
{
  "relationships": {"author": {}, "comments": {}}
}
```

An empty relationships member is not invalid, but kind of silly:

```
{
  "relationships": {}
}
```

The following is valid:

```
{
  "relationships": {"author": {"data": null}, "comments": {"data": []}}
}
```

And so is this:

```
{
  "type": "articles",
  "id": "1",
  "attributes": {
    "title": "Rails is Omakase"
  },
  "relationships": {
    "author": {
      "links": {
        "self": "/articles/1/relationships/author",
        "related": "/articles/1/author"
      },
      "data": { "type": "people", "id": "9" }
    }
  }
}
```

Note that the `data` element above is totally optional. It's not required, except in the case of a compound document.

jo-tham commented:
Thanks, @dgeb 
I think @sloria and I each understood half of this ;)

@sloria, anything else I need to add for this PR?

sloria commented:
Thanks @dgeb . 

@jo-tham We need tests for serializing null/empty relationships when `include_data=False`, which is the use case I was trying to clarify above.

jo-tham commented:
Would this do?

``` python
    def test_exclude_data(self, post_with_null_comment):
        field = Relationship(
            related_url='/posts/{post_id}/comments',
            related_url_kwargs={'post_id': '<id>'},
            many=True, include_data=False, type_='comments'
        )
        result = field.serialize('comments', post_with_null_comment)
        assert result['comments'] and result['comments']['links']['related']
        with pytest.raises(KeyError):
            result['comments']['data']

```

I suppose it doesn't really matter whether the relationship is empty - if empty or populated, we are excluding `data` object

sloria commented:
@jo-tham Yes, that looks good. One minor suggestion is to do

``` python
assert 'data' not in result['comments']
```

To make it clear that we're making an assertion about `'data'`, not `'comments'`.

jo-tham commented:
Good call :)

I have updated and removed "WIP:" from title.

sloria commented:
Looks good. Thanks!
